src/tokenserver/auth/browserid.rs

```diff
@@ -0,0 +1,486 @@
+            let response_body = response
+                .json::<VerifyResponse>()
+                .map_err(|_| TokenserverError::resource_unavailable())?;
+
+            match response_body {
+                VerifyResponse::Failure {
+                    reason: Some(reason),
+                } if reason.contains("expired") || reason.contains("issued later than") => {
+                    Err(TokenserverError {
+                        status: "invalid-timestamp",
+                        location: ErrorLocation::Body,
+                        ..Default::default()
+                    })
+                }
+                VerifyResponse::Failure { .. } => {
+                    Err(TokenserverError::invalid_credentials("Unauthorized"))
+                }
+                VerifyResponse::Okay { issuer, .. } if issuer != verifier.issuer => {
+                    Err(TokenserverError::invalid_credentials("Unauthorized"))
+                }
+                VerifyResponse::Okay {
+                    idp_claims: Some(claims),
+                    ..
+                } if !claims.token_verified() => {
+                    Err(TokenserverError::invalid_credentials("Unauthorized"))
+                }
+                VerifyResponse::Okay {
+                    email,
+                    idp_claims: Some(claims),
+                    ..
+                } => Ok(VerifyOutput {
+                    device_id: claims.device_id.clone(),
+                    email,
+                    generation: claims.generation()?,
+                    keys_changed_at: claims.keys_changed_at()?,
+                }),
+                VerifyResponse::Okay { email, .. } => Ok(VerifyOutput {
+                    device_id: None,
+                    email,
+                    generation: None,
+                    keys_changed_at: None,
+                }),
+            }
+        })
+        .await
+        .map_err(Into::into)
+    }
+}
+
+/// The request sent to the FxA BrowserID verifier for token verification.
+#[derive(Serialize)]
+struct VerifyRequest {
+    assertion: String,
+    audience: String,
+    #[serde(rename(serialize = "trustedIssuers"))]
+    trusted_issuers: [String; 1],
+}
+
+/// The response returned by the FxA BrowserID verifier for a token verification request.
+#[derive(Deserialize, Serialize)]
+#[serde(tag = "status", rename_all = "lowercase")]
+enum VerifyResponse {
+    Okay {
+        email: String,
+        #[serde(rename = "idpClaims")]
+        idp_claims: Option<IdpClaims>,
+        issuer: String,
+    },
+    Failure {
+        reason: Option<String>,
+    },
+}
+
+/// The claims extracted from a valid BrowserID assertion.
+#[derive(Deserialize, Serialize)]
+struct IdpClaims {
+    #[serde(rename = "fxa-deviceId")]
+    pub device_id: Option<String>,
+    #[serde(
+        default,
+        rename = "fxa-generation",
+        deserialize_with = "strict_deserialize"
+    )]
+    generation: Option<Option<i64>>,
```

Comment:
Related to [this thread](https://github.com/mozilla-services/syncstorage-rs/pull/1216#discussion_r799818023)

The second `Option` indicates whether the field was present or not in the JSON returned by FxA. I chose to use an `Option` here instead of a `Result` because it's not actually an error if one of these fields is missing in the response